vis_helpers/pca.py

```diff
@@ -3,104 +3,156 @@
         components_num = st.slider('Principal components', 1, 3, value=2)
         components_num = int(components_num)
-
+    
+    print_widgets_separator(1, sidebar=False)
+    
     cols = st.beta_columns(groups_num)
     dfs = []
+    group_names = []
     for col_ind, col in enumerate(cols):
         with col:
             files = st.file_uploader(label='Upload data',
                                      accept_multiple_files=True,
                                      type=['txt', 'csv'],
                                      key=f'uploader_{col_ind}'
                                      )
+            
+            spectrometer = st.selectbox("Choose spectra type",
+                                        spectra_types,
+                                        format_func=LABELS.get,
+                                        index=1,
+                                        key=col
+                                        )
+            group_name = st.text_input("Provide group name", f"{col_ind}")
+            group_names.append(group_name)
+            
             if files:
                 df = processing.save_read.files_to_df(files, spectrometer)
                 dfs.append(df)
-
+    
+    print_widgets_separator(1, sidebar=False)
+    
     if len(dfs) < groups_num:
         return
-
+    
     df = pd.concat(dfs, axis=1)
     df = df.interpolate().bfill().ffill()
-    group = [str(i) for i in range(groups_num) for _ in dfs[i].columns]
-
+    group = [group_names[i] for i in range(groups_num) for _ in dfs[i].columns]
+    
     if rescale:
         scaler = MinMaxScaler()
         rescaled_data = scaler.fit_transform(df)
         df = pd.DataFrame(rescaled_data, columns=df.columns, index=df.index)
-
-    fig = px.line(df, x=df.index, y=df.columns)
-    fig_layout(plot_template, fig, plots_colorscale=plot_palette)
-    for trace, gr in zip(fig.data, group):
-        trace['line']['color'] = plot_palette[int(gr)]
-    st.plotly_chart(fig, use_container_width=True)
-
+    
+    if baseline_correction:
+        baseline_corr_properties = st.sidebar.beta_expander('Change Baseline Correction Properties')
+        with baseline_corr_properties:
+            deg = utils.choosing_regression_degree()
+            window = utils.choosing_smoothening_window()
+            vals = {'col': (deg, window)}
+        
+        baselines = pd.DataFrame(index=df.index)
+        baselined = pd.DataFrame(index=df.index)
+        flattened = pd.DataFrame(index=df.index)
+        
+        for col in df.columns:
+            tmp_spectrum = df[col].dropna()  # trick for data with NaNs
+            tmp_spectrum = pd.Series(peakutils.baseline(tmp_spectrum, vals['col'][0]), index=tmp_spectrum.index)
+            baselines[col] = tmp_spectrum
+            
+            baselined[col] = df[col] - baselines[col]
+            flattened[col] = baselined[col].rolling(window=vals['col'][1], min_periods=1, center=True).mean()
+        df = flattened
```

Comment:
To jest skopiowany kawałek kodu z vis.py
Jak chcesz z tego korzystać w >1 miejscu to zrób z tego funkcję

Reply:
done